File: 0962-maximum-width-ramp/0962-maximum-width-ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        int len = nums.size();
        stack<int> st;
        for (int i = 0; i < len; ++i) {
            if (st.empty() || nums[st.top()] > nums[i]) st.push(i);
        }
        int res = 0;
        for (int i = len - 1; i; --i) {
            while (!st.empty() && nums[st.top()] <= nums[i]) {
                res = max(res, i - st.top());
                st.pop();
            }
            if (st.empty()) break;
        }
        return res;
    }
};

File: 0962-maximum-width-ramp/0962-maximum-width-ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int len=nums.length;
        Deque<Integer> st = new ArrayDeque<>();
        for (int i = 0; i < len; ++i) {
            if (st.isEmpty() || nums[st.peek()] > nums[i]) {
                st.push(i);
            }
        }
        int res = 0;
        for (int i = len - 1; i >= 0; --i) {
            while (!st.isEmpty() && nums[st.peek()] <= nums[i]) {
                res = Math.max(res, i - st.pop());
            }
            if (st.isEmpty()) {
                break;
            }
        }
        return res;

    }
}